sentinel-extension/sentinel-datasource-opensergo/src/main/java/com/alibaba/csp/sentinel/datasource/OpenSergoRuleAggregator.java

```diff
@@ -0,0 +1,303 @@
+    private volatile Map<String, ConcurrencyLimitStrategy> concurrencyLimitStrategyMap = new HashMap<>();
+    private volatile Map<String, CircuitBreakerStrategy> circuitBreakerStrategyMap = new HashMap<>();
+
+    public synchronized boolean updateFaultToleranceRuleList(List<FaultToleranceRule> rules) {
+        Map<String, List<FaultToleranceRule>> map = new HashMap<>(4);
+
+        if (rules != null && !rules.isEmpty()) {
+            for (FaultToleranceRule rule : rules) {
+                Set<String> kinds = rule.getStrategiesList().stream()
+                    .map(e -> e.getKind()).collect(Collectors.toSet());
+                kinds.forEach(kindName -> map.computeIfAbsent(kindName, v -> new ArrayList<>()).add(rule));
+            }
+        }
+        this.ftRuleMapByStrategyKind = map;
+
+        // TODO: check whether the rules have been changed
+        handleFlowRuleUpdate();
+        handleCircuitBreakerRuleUpdate();
+
+        return true;
+    }
+
+    public synchronized boolean updateRateLimitStrategy(List<RateLimitStrategy> strategies) {
+        Map<String, RateLimitStrategy> map = new HashMap<>(4);
+        if (strategies != null && !strategies.isEmpty()) {
+            strategies.forEach(s -> map.put(s.getName(), s));
+        }
+        this.rateLimitStrategyMap = map;
+
+        return handleFlowRuleUpdate();
+    }
+
+    public synchronized boolean updateThrottlingStrategy(List<ThrottlingStrategy> strategies) {
+        Map<String, ThrottlingStrategy> map = new HashMap<>(4);
+        if (strategies != null && !strategies.isEmpty()) {
+            strategies.forEach(s -> map.put(s.getName(), s));
+        }
+        this.throttlingStrategyMap = map;
+
+        return handleFlowRuleUpdate();
+    }
+
+    public synchronized boolean updateCircuitBreakerStrategy(List<CircuitBreakerStrategy> strategies) {
+        Map<String, CircuitBreakerStrategy> map = new HashMap<>(4);
+        if (strategies != null && !strategies.isEmpty()) {
+            strategies.forEach(s -> map.put(s.getName(), s));
+        }
+        this.circuitBreakerStrategyMap = map;
+
+        return handleCircuitBreakerRuleUpdate();
+    }
+
+    private boolean handleFlowRuleUpdate() {
+        List<FlowRule> rules = new ArrayList<>();
+
+        List<FlowRule> rulesFromRateLimitStrategies = assembleFlowRulesFromRateLimitStrategies(
+            ftRuleMapByStrategyKind.get(ConfigKind.RATE_LIMIT_STRATEGY.getSimpleKindName()), rateLimitStrategyMap);
+        List<FlowRule> rulesFromThrottlingStrategies = assembleFlowRulesFromThrottlingStrategies(
+            ftRuleMapByStrategyKind.get(ConfigKind.THROTTLING_STRATEGY.getSimpleKindName()), throttlingStrategyMap);
+
+        rules.addAll(rulesFromRateLimitStrategies);
+        rules.addAll(rulesFromThrottlingStrategies);
+
+        // Update rules to upstream data-source.
+        return dataSourceMap.get(OpenSergoSentinelConstants.KIND_FLOW_RULE).updateValue(rules);
+    }
```

Comment:
maybe lost rules by update new-datas override all the older-datas:
 if there were 10 rules in OpenSergo Control Plane.
1、in the 1st time, sentinel will load all datas from opensergo, the number is 10.
2、if 4 rules of the 10 in OpenSergo Control Plane were changed, the other 6 had no changes, opensergo will push the 4 rules to sentinel. And sentinel will load them 2nd time, but the number is 4.

In current logic, the 4 rules will override the 10 rules which were loaded into Sentinel in the 1st time. The 6 rules which had no changed would be lost.

So I think, shoud assmeble the whole 10 rules by update 4 rules before invoke UpdateValue.